<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Read top-down files. — readTopDownFiles • topdownr</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">topdownr</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/analysis.html">Fragmentation Analysis with topdownr</a>
    </li>
    <li>
      <a href="../articles/data-generation.html">Data Generation for topdownr</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/sgibb/topdownr/">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Read top-down files.</h1>
    </div>

    
    <p>It creates an <a href='TopDownSet-class.html'>TopDownSet</a> object and is its only constructor.</p>
    

    <pre class="usage"><span class='fu'>readTopDownFiles</span>(<span class='no'>path</span>, <span class='kw'>pattern</span> <span class='kw'>=</span> <span class='st'>".*"</span>, <span class='kw'>type</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"a"</span>, <span class='st'>"b"</span>, <span class='st'>"c"</span>, <span class='st'>"x"</span>, <span class='st'>"y"</span>,
  <span class='st'>"z"</span>), <span class='kw'>modifications</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"Carbamidomethyl"</span>, <span class='st'>"Acetyl"</span>, <span class='st'>"Met-loss"</span>),
  <span class='kw'>adducts</span> <span class='kw'>=</span> <span class='fu'>data.frame</span>(), <span class='kw'>neutralLoss</span> <span class='kw'>=</span> <span class='kw pkg'>MSnbase</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/MSnbase/topics/calculateFragments-methods'>defaultNeutralLoss</a></span>(),
  <span class='kw'>sequenceOrder</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"original"</span>, <span class='st'>"random"</span>, <span class='st'>"inverse"</span>), <span class='kw'>tolerance</span> <span class='kw'>=</span> <span class='fl'>5e-06</span>,
  <span class='kw'>dropNonInformativeColumns</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>sampleColumns</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"Mz"</span>, <span class='st'>"AgcTarget"</span>,
  <span class='st'>"EtdReagentTarget"</span>, <span class='st'>"EtdActivation"</span>, <span class='st'>"CidActivation"</span>, <span class='st'>"HcdActivation"</span>,
  <span class='st'>"UvpdActivation"</span>), <span class='kw'>verbose</span> <span class='kw'>=</span> <span class='fu'>interactive</span>())</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a> Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>path</th>
      <td><p><code>character</code>,
path to directory that contains the top-down files.</p></td>
    </tr>
    <tr>
      <th>pattern</th>
      <td><p><code>character</code>,
a filename pattern, the default <code>.*</code> means all files.</p></td>
    </tr>
    <tr>
      <th>type</th>
      <td><p><code>character</code>,
type of fragments, currently <em>a-c</em> and <em>x-z</em> are
supported, see
<code><a href='http://www.rdocumentation.org/packages/MSnbase/topics/calculateFragments-methods'>MSnbase::calculateFragments()</a></code>
for details.</p></td>
    </tr>
    <tr>
      <th>modifications</th>
      <td><p><code>character</code>,
unimod names of modifications that should be applied.
Currenlty just <em>Acetyl</em> (Unimod:1 but just protein N-term),
<em>Carbamidomethyl</em> (Unimod:4) and
<em>Met-loss</em> (Unimod:765) are supported.
<em>Met-loss</em> removes M
(if followed by A, C, G, P, S, T, or V;
(see also
http://www.unimod.org/modifications_view.php?editid1=1,
http://www.unimod.org/modifications_view.php?editid1=4, and
http://www.unimod.org/modifications_view.php?editid1=765 for details)).
Use <code>NULL</code> to disable all modifications.</p></td>
    </tr>
    <tr>
      <th>adducts</th>
      <td><p><code>data.frame</code>,
with 3 columns, namely: mass, name, to, see details section.</p></td>
    </tr>
    <tr>
      <th>neutralLoss</th>
      <td><p><code>list</code>,
neutral loss that should be applied, see
<code><a href='http://www.rdocumentation.org/packages/MSnbase/topics/calculateFragments-methods'>MSnbase::calculateFragments()</a></code> and
<code><a href='http://www.rdocumentation.org/packages/MSnbase/topics/calculateFragments-methods'>MSnbase::defaultNeutralLoss()</a></code>
for details.</p></td>
    </tr>
    <tr>
      <th>sequenceOrder</th>
      <td><p><code>character</code>,
order of the sequence before fragment calculation and matching is done.
<code>"original"</code> doesn't change anything.
<code>"inverse"</code> reverse the sequence and
<code>"random"</code> arranges the amino acid sequence at random.</p></td>
    </tr>
    <tr>
      <th>tolerance</th>
      <td><p><code>double</code>,
tolerance in <em>ppm</em> that is used to match the
theoretical fragments with the observed ones.</p></td>
    </tr>
    <tr>
      <th>dropNonInformativeColumns</th>
      <td><p>logical,
should columns with just one identical value across all runs be removed?</p></td>
    </tr>
    <tr>
      <th>sampleColumns</th>
      <td><p><code>character</code>,
column names of the <code><a href='AbstractTopDownSet-class.html'>colData()</a></code>
used to define a sample (technical replicate). This is used to add the
<code>Sample</code> column (used for easier aggregation, etc.).</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p><code>logical</code>, verbose output?</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <code>TopDownSet</code> object.</p>
    
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p><code>readTopDownFiles</code> reads and processes all top-down files, namely:</p><ul>
<li><p><code>.fasta</code> (protein sequence)</p></li>
<li><p><code>.mzML</code> (spectra)</p></li>
<li><p><code>.experiments.csv</code> (method/fragmentation conditions)</p></li>
<li><p><code>.txt</code> (scan header information)</p></li>
</ul><p><code>adducts</code>: <em>Thermo's Xtract</em>
allows some mistakes in deisotoping, mostly it
allows <code>+/- C13-C12</code> and <code>+/- H+</code>.
The <code>adducts</code> argument takes a
<code>data.frame</code> with the <code>mass</code> to add, the <code>name</code>
that should assign to these
new fragments and an information <code>to</code>
whom the modification should be
applied, e.g. for <code>H+</code> on <code>z</code>,
<code>data.frame(mass=1.008, name="zpH", to="z")</code>.
    <em>Please note:</em> The <code>adducts</code> are added to the output of
<code><a href='http://www.rdocumentation.org/packages/MSnbase/topics/calculateFragments-methods'>MSnbase::calculateFragments()</a></code>.
That has some limitations, e.g.
neutral loss calculation could not be done in
<a href='topdownr-package.html'>topdownr-package</a>.
If neutral loss should be applied on adducts you have to create
additional rows, e.g.:
<code>data.frame(mass=c(1.008, 1.008), name=c("cpH", "cpH_"), to=c("c", "c_"))</code>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <p><code><a href='http://www.rdocumentation.org/packages/MSnbase/topics/calculateFragments-methods'>MSnbase::calculateFragments()</a></code>,
<code><a href='http://www.rdocumentation.org/packages/MSnbase/topics/calculateFragments-methods'>MSnbase::defaultNeutralLoss()</a></code></p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='kw'>if</span> (<span class='fu'>require</span>(<span class='st'>"topdownrdata"</span>)) {
    <span class='co'># add H+ to z and no neutral loss of water</span>
    <span class='no'>tds</span> <span class='kw'>&lt;-</span> <span class='fu'>readTopDownFiles</span>(
        <span class='kw pkg'>topdownrdata</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/topdownrdata/topics/topDownDataPath'>topDownDataPath</a></span>(<span class='st'>"myoglobin"</span>),
        <span class='co'>## Use an artifical pattern to load just the fasta</span>
        <span class='co'>## file and files from m/z == 1211, ETD reagent</span>
        <span class='co'>## target 1e6 and first replicate to keep runtime</span>
        <span class='co'>## of the example short</span>
        <span class='kw'>pattern</span><span class='kw'>=</span><span class='st'>".*fasta.gz$|1211_.*1e6_1"</span>,
        <span class='kw'>adducts</span><span class='kw'>=</span><span class='fu'>data.frame</span>(<span class='kw'>mass</span><span class='kw'>=</span><span class='fl'>1.008</span>, <span class='kw'>name</span><span class='kw'>=</span><span class='st'>"zpH"</span>, <span class='kw'>to</span><span class='kw'>=</span><span class='st'>"z"</span>),
        <span class='kw'>neutralLoss</span><span class='kw'>=</span><span class='kw pkg'>MSnbase</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/MSnbase/topics/calculateFragments-methods'>defaultNeutralLoss</a></span>(
            <span class='kw'>disableWaterLoss</span><span class='kw'>=</span><span class='fu'>c</span>(<span class='st'>"Cterm"</span>, <span class='st'>"D"</span>, <span class='st'>"E"</span>, <span class='st'>"S"</span>, <span class='st'>"T"</span>)),
        <span class='kw'>tolerance</span><span class='kw'>=</span><span class='fl'>25e-6</span>
   )
}</div><div class='output co'>#&gt; <span class='message'>Loading required package: topdownrdata</span></div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#details">Details</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Sebastian Gibb, Pavel Shliaha, Ole Nørregaard Jensen.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
